File: solutions/java/square-root/1/src/main/java/SquareRoot.java
public class SquareRoot {
    public int squareRoot(int radicand) {
        return binarySearch(radicand, 0, radicand + 1);
    }

    private static int binarySearch(int radicand, int left, int right) {
        while(left != right - 1) {
            final int middle = (left + right) >> 1;

            if (middle * middle <= radicand)
                left = middle;
            else
                right = middle;
        }
        
        return left;
    }
}

File: solutions/elixir/square-root/1/lib/square_root.ex
defmodule SquareRoot do
  @doc """
  Calculate the integer square root of a positive integer
  """
  @spec calculate(radicand :: pos_integer) :: pos_integer
  def calculate(radicand), do: binary_search(radicand, 0, radicand + 1)

  defp binary_search(radicand, left, right) do
    middle = div(left + right, 2)

    cond do
      left == right - 1 -> left
      middle * middle <= radicand -> binary_search(radicand, middle, right)
      :else -> binary_search(radicand, left, middle)
    end
  end
end
